### ISSUE TITLE
With `--use-amp`, the FGL/DDF losses run in fp16 (criterion under `autocast(enabled=False)` on fp16 outputs) — DDF reads ≈ 5× its fp32 value

### ISSUE THREAD

**Issue by peymannaderi10** [Peterande/D-FINE#361]
`src/solver/det_engine.py::train_one_epoch` computes the forward under `torch.autocast(...)` and then the criterion
under `torch.autocast(device_type=..., enabled=False)`:

```python
with torch.autocast(device_type=str(device), cache_enabled=True):
    outputs = model(samples, targets=targets)
...
with torch.autocast(device_type=str(device), enabled=False):
    loss_dict = criterion(outputs, targets, **metas)
```

`outputs["pred_corners"]` / `["teacher_corners"]` are fp16 tensors under autocast, and `DFINECriterion.loss_local`
uses them as they are: `F.cross_entropy(pred, ...)` in `unimodal_distribution_focal_loss` (FGL) and
`F.kl_div(F.log_softmax(pred_corners / T, dim=1), F.softmax(target_corners.detach() / T, dim=1), ...)` (DDF).
With autocast *disabled* around the criterion these ops run in the input dtype (fp16) — autocast would otherwise
promote `softmax` / `log_softmax` / `cross_entropy` / `kl_div` to fp32. `weight.float()` (line ~509) only casts
the weights.

Measured on the same checkpoint (epoch 35 of a 42-epoch fine-tune; `-r`, same seed and data order), loss-family sums
over all layers at step 200–300 of the next epoch:

| numerics | total | loss_fgl | loss_vfl | loss_giou | loss_bbox | loss_ddf |
|---|---:|---:|---:|---:|---:|---:|
| `--use-amp` (fp16 forward, criterion on fp16 outputs — the README command) | 34.5 | 12.2 | 6.4 | 6.0 | 1.3 | **8.4** |
| bf16 autocast forward, outputs cast to fp32 before the criterion | 27.8 | 11.9 | 6.3 | 6.0 | 1.3 | 1.4 |
| no autocast (fp32) | 27.6 | 11.9 | 6.3 | 6.1 | 1.4 | **1.9** |

Every family agrees except DDF, which is ≈ 5× larger in fp16. (bf16 autocast with the criterion on *bf16*
outputs is far worse: `loss_fgl` 44–47 and `loss_ddf` 11–12 per term, total 330–390 — 8 mantissa bits are not
enough for the distribution losses.) Separately, on this run the fp16 forward started producing NaN
`pred_boxes` in stage 2 (the `NaN.pth` dump, then the `generalized_box_iou` assertion in the matcher took one
rank down and the other timed out on NCCL); bf16 with fp32 loss inputs runs clean.

Suggested fix: cast the outputs used by the criterion to fp32 (e.g. `outputs = {k: v.float() ...}` recursively,
or `.float()` on `pred_corners` / `teacher_corners` / logits inside `loss_local` and `loss_labels_vfl`), or drop
the `enabled=False` wrapper so autocast's fp32 promotion applies. Happy to open a PR; the change we run is a
15-line patch on `det_engine.py`.

Environment: torch 2.7 (NGC PyTorch 25.02 container), CUDA 12.8, Ampere GPUs; also reproduced on the same commit with the
D-FINE-L Objects365+COCO weights as `-t`.
